cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+			    NULL, 0, NI_NUMERICHOST);
+			if (s != 0) {
+				ZREPL_ERRLOG("getnameinfo() failed: %d\n",
+				    errno);
+				s = -1;
+				goto exit;
+			}
+
+			if (family == AF_INET) {
+				if (strcmp(host, "127.0.0.1") == 0) {
+					continue;
+				}
+				printf("IP address: %s\n", host);
+				break;
+			}
+		}
+	}
+exit:
+	freeifaddrs(ifaddr);
+	return (s);
+}
+/*
+ * Allocate zio command along with
+ * buffer needed for IO completion.
+ */
+static zvol_io_cmd_t *
+zio_cmd_alloc(zvol_io_hdr_t *hdr, int fd)
+{
+	zvol_io_cmd_t *zio_cmd = kmem_zalloc(
+	    sizeof (zvol_io_cmd_t), KM_SLEEP);
+
+	bcopy(hdr, &zio_cmd->hdr, sizeof (zio_cmd->hdr));
+	if ((hdr->opcode == ZVOL_OPCODE_READ) ||
+	    (hdr->opcode == ZVOL_OPCODE_WRITE) ||
+	    (hdr->opcode == ZVOL_OPCODE_HANDSHAKE)) {
+		zio_cmd->buf = kmem_zalloc(sizeof (char) * hdr->len, KM_SLEEP);
+	}
+
+	zio_cmd->conn = fd;
+	return (zio_cmd);
+}
+
+/*
+ * Free zio command along with buffer.
+ */
+static void
+zio_cmd_free(zvol_io_cmd_t **cmd, zvol_op_code_t opcode)
+{
+	zvol_io_cmd_t *zio_cmd = *cmd;
+	switch (opcode) {
+		case ZVOL_OPCODE_READ:
+		case ZVOL_OPCODE_WRITE:
+		case ZVOL_OPCODE_HANDSHAKE:
+			if (zio_cmd->buf != NULL) {
+				free(zio_cmd->buf);
+			}
+			break;
+		default:
+			ASSERT(!"Wrong Op code");
+			break;
+	}
+
+	free(zio_cmd);
+	*cmd = NULL;
+}
+
+
+static int
+uzfs_zvol_socket_read(int fd, char *buf, uint64_t nbytes)
+{
+	uint64_t count = 0;
+	char *p = buf;
+	ZREPL_ERRLOG("Trying to read nbytes: %lu\n", nbytes);
+	while (nbytes) {
+		count = read(fd, (void *)p, nbytes);
+		if ((count <= 0) && (errno == EAGAIN)) {
+			continue;
+		} else if (count <= 0) {
+			printf("Read error\n");
+			return (-1);
+		}
+
+		ZREPL_ERRLOG("In read count:%lu nbytes: %lu\n", count, nbytes);
+		p += count;
+		nbytes -= count;
+	}
+	ZREPL_LOG("Successful read count:%lu nbytes: %lu\n", count, nbytes);
+	return (1);
+}
+
+
+static inline int
+uzfs_zvol_socket_write(int fd, char *buf, int nbytes)
+{
+	int count = 0;
+	char *p = buf;
+	while (nbytes) {
+		count = write(fd, (void *)p, nbytes);
+		if (count <= 0) {
+			printf("Write error\n");
+			return (-1);
+		}
+		p += count;
+		nbytes -= count;
+	}
+	return (0);
+}
+
+/*
+ * zvol worker is responsible for actual work.
+ * It execute read/write/sync command to uzfs.
+ * It enqueue command to completion queue and
+ * send signal to ack-sender thread.
+ */
+static void
+uzfs_zvol_worker(void *arg)
+{
+	zvol_io_cmd_t	*zio_cmd;
+	zvol_info_t	*zinfo;
+	zvol_io_hdr_t 	*hdr;
+	int		rc = 0;
+	int 		write = 0;
+
+
+	zio_cmd = (zvol_io_cmd_t *)arg;
+	hdr = &zio_cmd->hdr;
+	zinfo = zio_cmd->zv;
+	ASSERT(zinfo);
+	switch (hdr->opcode) {
+		case ZVOL_OPCODE_READ:
+			rc = uzfs_read_data(zinfo->zv,
+			    (char *)zio_cmd->buf,
+			    hdr->offset, hdr->len, NULL, NULL);
+			break;
+
+		case ZVOL_OPCODE_WRITE:
+			write = 1;
+			rc = uzfs_write_data(zinfo->zv,
+			    (char *)zio_cmd->buf,
+			    hdr->offset, hdr->len, NULL);
+			break;
+
+		case ZVOL_OPCODE_SYNC:
```

Comment:
Vitta, has provided zil_commit API very recently. I will call that API here. This was not added because I need to add SYNC command in Replica which I deferred for my next sprint.

Reply:
👍